src/go/networking/pkg/portproxy/server.go

```diff
@@ -73,47 +83,139 @@ func (p *PortProxy) handleEvent(conn net.Conn) {
-		for _, portBinding := range portBindings {
-			logrus.Debugf("received the following port: [%s] from portMapping: %+v", portBinding.HostPort, pm)
-			port, err := nat.ParsePort(portBinding.HostPort)
-			if err != nil {
-				logrus.Errorf("parsing port error: %s", err)
-				continue
-			}
-			if pm.Remove {
-				p.mutex.Lock()
-				if listener, exist := p.activeListeners[port]; exist {
-					logrus.Debugf("closing listener for port: %d", port)
-					if err := listener.Close(); err != nil {
-						logrus.Errorf("error closing listener for port [%s]: %s", portBinding.HostPort, err)
-					}
+func (p *PortProxy) exec(pm types.PortMapping) {
+	for portProto, portBindings := range pm.Ports {
+		logrus.Debugf("received the following port: [%s] and protocol: [%s] from portMapping: %+v", portProto.Port(), portProto.Proto(), pm)
+		if gvisorTypes.TransportProtocol(portProto.Proto()) == gvisorTypes.TCP {
+			p.handleTCP(portBindings, pm.Remove)
+		}
+		if gvisorTypes.TransportProtocol(portProto.Proto()) == gvisorTypes.UDP {
+			p.handleUDP(portBindings, pm.Remove)
+		}
+	}
+}
+
+func (p *PortProxy) handleUDP(portBindings []nat.PortBinding, remove bool) {
+	for _, portBinding := range portBindings {
+		port, err := nat.ParsePort(portBinding.HostPort)
+		if err != nil {
+			logrus.Errorf("parsing port error: %s", err)
+			continue
+		}
+		if remove {
+			p.udpConnMutex.Lock()
+			if udpConn, exist := p.activeUDPConns[port]; exist {
+				logrus.Debugf("closing UDPConn for port: %d", port)
+				if err := udpConn.Close(); err != nil {
+					logrus.Errorf("error closing UDPConn for port [%s]: %s", portBinding.HostPort, err)
 				}
-				delete(p.activeListeners, port)
-				p.mutex.Unlock()
-				continue
 			}
-			addr := net.JoinHostPort(portBinding.HostIP, portBinding.HostPort)
-			l, err := net.Listen("tcp", addr)
-			if err != nil {
-				logrus.Errorf("failed creating listener for published port [%s]: %s", portBinding.HostPort, err)
-				continue
+			delete(p.activeUDPConns, port)
+			p.udpConnMutex.Unlock()
+			continue
+		}
+
+		// the localAddress IP section can either be 0.0.0.0 or 127.0.0.1
+		localAddress := net.JoinHostPort(portBinding.HostIP, portBinding.HostPort)
+		sourceAddr, err := net.ResolveUDPAddr("udp", localAddress)
+		if err != nil {
+			logrus.Errorf("failed to resolve UDP source address [%s]: %s", sourceAddr, err)
+			continue
+		}
+
+		c, err := net.ListenUDP("udp", sourceAddr)
+		if err != nil {
+			logrus.Errorf("failed creating listener for published port [%s]: %s", portBinding.HostPort, err)
+			continue
+		}
+
+		p.udpConnMutex.Lock()
+		defer p.udpConnMutex.Unlock()
+		p.activeUDPConns[port] = c
+		logrus.Debugf("created UDPConn for: %v", sourceAddr)
+
+		forwardAddr := net.JoinHostPort(p.config.UpstreamAddress, portBinding.HostPort)
+		targetAddr, err := net.ResolveUDPAddr("udp", forwardAddr)
+		if err != nil {
+			logrus.Errorf("failed to resolve UDP target address [%s]: %s", targetAddr, err)
```

Comment:
Do we need to close `c` in this case? It's not going anywhere… (Shouldn't we do this set up before listening?)

Reply:
Yes, that would be better.